ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/PerfRouter.java

```diff
@@ -0,0 +1,291 @@
+    } else {
+      completeOperation(futureResult, callback, null, null);
+    }
+    return futureResult;
+  }
+
+  @Override
+  public void close() {
+    try {
+      if (routerOpen.compareAndSet(true, false)) {
+        operationPool.shutdown();
+        operationPool.awaitTermination(1, TimeUnit.MINUTES);
+      } else {
+        operationPool.awaitTermination(1, TimeUnit.MINUTES);
+      }
+      logger.info("PerfRouter closed");
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Completes a router operation by invoking the {@code callback} and setting the {@code futureResult} with
+   * {@code operationResult} (if any) and {@code exception} (if any).
+   * @param futureResult the {@link FutureResult} that needs to be set.
+   * @param callback that {@link Callback} that needs to be invoked. Can be null.
+   * @param operationResult the result of the operation (if any).
+   * @param exception {@link Exception} encountered while performing the operation (if any).
+   * @param <T> the type of {@code futureResult}, {@code callback} and {@code operationResult}.
+   */
+  private <T> void completeOperation(FutureResult<T> futureResult, Callback<T> callback, T operationResult,
+      Exception exception) {
+    RuntimeException runtimeException = null;
+    if (exception != null) {
+      runtimeException = new RuntimeException(exception);
+    }
+    futureResult.done(operationResult, runtimeException);
+    if (callback != null) {
+      callback.onCompletion(operationResult, exception);
+    }
+  }
+
+  /**
+   * Gets a random string of size {@code length}.
+   * @param length the size of the required random string.
+   * @return a random string of size {@code length}.
+   */
+  private String getRandomString(int length) {
+    StringBuilder sb = new StringBuilder(length);
+    for (int i = 0; i < length; i++) {
+      sb.append(CHARACTERS.charAt(random.nextInt(CHARACTERS.length())));
+    }
+    return sb.toString();
+  }
+
+  /**
+   * Thread that consumes put content async.
+   */
+  private class PutContentConsumer implements Runnable {
+    private final ReadableStreamChannel putContentChannel;
+    private final FutureResult<String> futureResult;
+    private final Callback<String> callback;
+    private final long queueStartTime = System.currentTimeMillis();
+
+    /**
+     * Create a new instance.
+     * @param putContentChannel the {@link ReadableStreamChannel} on which put content is available.
+     * @param futureResult the {@link FutureResult} that must be set on completion.
+     * @param callback the {@link Callback} to be invoked on completion.
+     */
+    public PutContentConsumer(ReadableStreamChannel putContentChannel, FutureResult<String> futureResult,
+        Callback<String> callback) {
+      this.putContentChannel = putContentChannel;
+      this.futureResult = futureResult;
+      this.callback = callback;
+    }
+
+    @Override
+    public void run() {
```

Comment:
When `NoOpAWC` did not have a thread, this was required because even though everything is non blocking it still takes a lot of time. 

Copying data in-memory is non-blocking too but we never do it in a call that is supposed to be async because the call is expected to be quick.